File: backend/utils/ephemeris.js
import swe from "swisseph";

const PLANETS = {
  Sun: swe.SUN,
  Moon: swe.MOON,
  Mars: swe.MARS,
  Mercury: swe.MERCURY,
  Jupiter: swe.JUPITER,
  Venus: swe.VENUS,
  Saturn: swe.SATURN
};

// Convert date → Julian day
function getJulianDay(date) {
  return swe.swe_julday(
    date.getFullYear(),
    date.getMonth() + 1,
    date.getDate(),
    date.getHours()
  );
}

// 🪐 REAL PLANET LONGITUDE CALCULATION
export function getPlanetPositions(dob, time) {
  const date = new Date(`${dob}T${time}`);
  const jd = getJulianDay(date);

  const result = {};

  Object.entries(PLANETS).forEach(([name, id]) => {
    const [longitude] = swe.swe_calc_ut(jd, id, swe.FLG_SWIEPH);
    result[name] = {
      longitude: longitude,
      sign: getZodiac(longitude)
    };
  });

  return result;
}

// 🔮 Convert longitude → zodiac sign
function getZodiac(deg) {
  const signs = [
    "Aries","Taurus","Gemini","Cancer",
    "Leo","Virgo","Libra","Scorpio",
    "Sagittarius","Capricorn","Aquarius","Pisces"
  ];
  return signs[Math.floor(deg / 30)];
}

File: backend/utils/ephemeris.py
import swisseph as swe
import datetime

# set ephemeris path (important)
swe.set_ephe_path('/usr/share/ephe')

PLANETS = {
    "Sun": swe.SUN,
    "Moon": swe.MOON,
    "Mars": swe.MARS,
    "Mercury": swe.MERCURY,
    "Jupiter": swe.JUPITER,
    "Venus": swe.VENUS,
    "Saturn": swe.SATURN,
    "Rahu": swe.MEAN_NODE,
    "Ketu": swe.TRUE_NODE
}

SIGNS = [
    "Aries","Taurus","Gemini","Cancer","Leo","Virgo",
    "Libra","Scorpio","Sagittarius","Capricorn","Aquarius","Pisces"
]

def get_zodiac(degree):
    return SIGNS[int(degree / 30)]

def get_planet_positions(date_str, time_str):
    dt = datetime.datetime.fromisoformat(f"{date_str}T{time_str}")

    jd = swe.julday(dt.year, dt.month, dt.day,
                    dt.hour + dt.minute/60)

    result = {}

    for name, planet in PLANETS.items():
        pos, _ = swe.calc_ut(jd, planet)

        longitude = pos[0]
        result[name] = {
            "degree": longitude,
            "sign": get_zodiac(longitude),
            "raw": pos
        }

    return result